lightning/src/ln/router.rs

```diff
@@ -1864,4 +1870,776 @@ mod tests {
+				node_signature_1: secp_ctx.sign(&msghash, node_1_privkey),
+				node_signature_2: secp_ctx.sign(&msghash, node_2_privkey),
+				bitcoin_signature_1: secp_ctx.sign(&msghash, node_1_btckey),
+				bitcoin_signature_2: secp_ctx.sign(&msghash, node_2_btckey),
+				contents: unsigned_announcement.clone(),
+			};
+			match router.handle_channel_announcement(&valid_channel_announcement) {
+				Ok(_) => (),
+				Err(_) => panic!()
+			};
+		}
+
+		// No updates.
+		let next_announcements = router.get_next_channel_announcements(channel_key, 2);
+		assert_eq!(next_announcements.len(), 0);
+
+		{
+			let unsigned_channel_update = UnsignedChannelUpdate {
+				chain_hash,
+				short_channel_id,
+				timestamp: 100,
+				flags: 0,
+				cltv_expiry_delta: 144,
+				htlc_minimum_msat: 1000000,
+				fee_base_msat: 10000,
+				fee_proportional_millionths: 20,
+				excess_data: Vec::new()
+			};
+			let msghash = hash_to_message!(&Sha256dHash::hash(&unsigned_channel_update.encode()[..])[..]);
+			let valid_channel_update = ChannelUpdate {
+				signature: secp_ctx.sign(&msghash, node_1_privkey),
+				contents: unsigned_channel_update
+			};
+			match router.handle_channel_update(&valid_channel_update) {
+				Ok(_) => (),
+				Err(_) => panic!()
+			};
+		}
+
+		// Updates with empty messages are not returned.
+		let next_announcements = router.get_next_channel_announcements(channel_key, 2);
+		assert_eq!(next_announcements.len(), 0);
+
+		{
+			// Necessary channel updates
+			let mut unsigned_channel_update = UnsignedChannelUpdate {
+				chain_hash,
+				short_channel_id,
+				timestamp: 101,
+				flags: 0,
+				cltv_expiry_delta: 144,
+				htlc_minimum_msat: 1000000,
+				fee_base_msat: 10000,
+				fee_proportional_millionths: 20,
+				excess_data: Vec::new()
+			};
+			let msghash = hash_to_message!(&Sha256dHash::hash(&unsigned_channel_update.encode()[..])[..]);
+			let valid_channel_update = ChannelUpdate {
+				signature: secp_ctx.sign(&msghash, node_1_privkey),
+				contents: unsigned_channel_update.clone()
+			};
+			match router.handle_channel_update(&valid_channel_update) {
+				Ok(_) => (),
+				Err(_) => panic!()
+			};
+
+			unsigned_channel_update.flags = 1;
+			let msghash = hash_to_message!(&Sha256dHash::hash(&unsigned_channel_update.encode()[..])[..]);
+			let valid_channel_update = ChannelUpdate {
+				signature: secp_ctx.sign(&msghash, node_2_privkey),
+				contents: unsigned_channel_update
+			};
+			match router.handle_channel_update(&valid_channel_update) {
+				Ok(_) => (),
+				Err(_) => panic!()
+			};
+		}
+
+		let next_announcements = router.get_next_channel_announcements(channel_key, 2);
+		assert_eq!(next_announcements.len(), 1);
+
+		// Further starting point have no channels after it
+		let next_announcements = router.get_next_channel_announcements(channel_key + 1000, 2);
+		assert_eq!(next_announcements.len(), 0);
+
+		{
+			//  An update with excess data, which will prevent us from returning any updates for a channel.
```

Comment:
Although I coded this test, I think this behavior is... strange. Are we sure it makes sense? Same for node_announcements too.

Reply:
The comment is confusing, as we aren't "returning" anything. We refuse to relay such announcements as they could very, very quickly grow pretty large and become a DoS issue both for our own RAM/disk as well as our bandwidth. We could limit the excess data length to something pretty tight (eg 100 bytes or so), but we definitely shouldn't have a worst-case which is much higher than our general case.

Reply:
I was referring to get_next_channel_announcements.
It returns 0 updates, even though *only last update* has excess data. So just ignoring all the previous ones.
That's what I was worrying about.

Reply:
Nvm, just my confusion, will delete.